llvm/lib/Target/AIE/AIEPostSelectOptimize.cpp

```diff
@@ -438,6 +446,120 @@ bool duplicateAdressingRegs(MachineBasicBlock &MBB, MachineRegisterInfo &MRI) {
   return tryToDuplicateLoadUse(LoadUses, NonLoadUses, MRI, TII);
 }
 
+bool getGlobalValue(const MachineInstr *MI,
+                    SmallPtrSet<const Value *, 4> &GVSet,
+                    MachineRegisterInfo &MRI) {
+
+  if (MI->isCopy())
+    MI = getDefIgnoringCopiesAndBitcasts(MI->getOperand(1).getReg(), MRI);
+
+  for (auto MO : MI->uses()) {
+    if (MO.isGlobal()) {
+      GVSet.insert(MO.getGlobal());
+      return true;
+    }
+  }
+  return false;
+}
+
+using Opcodes = AIEBaseInstrInfo::AbstractVecOp::AbstractOpcode;
+using AbstractVecOp = AIEBaseInstrInfo::AbstractVecOp;
+using OptionalVecOp = std::optional<AbstractVecOp>;
+
+/// This optimization tries to propagate GlobalValues as MMO of load operations
+/// when they are missing. This specific implementation considers only the
+/// following pattern:
+/// VOp (VSEL (VBCST(GlobalValue1), VBCST(GlobalValue2), Mask) Value)
+bool findBroadcastedGlobalValues(const MachineInstr *MI,
+                                 SmallPtrSet<const Value *, 4> &GVSet,
+                                 MachineRegisterInfo &MRI) {
+
+  const TargetSubtargetInfo &ST = MI->getParent()->getParent()->getSubtarget();
+  const AIEBaseInstrInfo *TII =
+      static_cast<const AIEBaseInstrInfo *>(ST.getInstrInfo());
+
+  SmallPtrSet<const MachineInstr *, 8> VisitedInsts;
+  SmallVector<const MachineInstr *, 8> WorkList;
+
+  WorkList.push_back(MI);
+
+  auto ParseMI = [&](const MachineInstr *MIToParse,
+                     Opcodes Opcode) -> OptionalVecOp {
+    if (!MIToParse)
+      return {};
+    OptionalVecOp VOp = TII->parseTargetVectorOp(*MIToParse);
+    if (VOp && VOp->Opcode == Opcode)
+      return VOp;
+    return {};
+  };
+
+  bool FoundSelect = false;
+  AbstractVecOp VSelect;
+  while (!FoundSelect && !WorkList.empty() &&
+         WorkList.size() < LoadMMOSearchLimit) {
+
+    const MachineInstr *CurrMI = WorkList.pop_back_val();
+    if (VisitedInsts.contains(CurrMI))
+      continue;
+    else
+      VisitedInsts.insert(CurrMI);
+
+    if (OptionalVecOp Op = ParseMI(CurrMI, Opcodes::SELECT)) {
+      VSelect = *Op;
+      FoundSelect = true;
+    } else {
+      for (auto MO : CurrMI->uses()) {
+        if (!MO.isReg() || !MO.getReg().isVirtual())
+          continue;
+        if (const MachineInstr *RegDef = MRI.getVRegDef(MO.getReg()))
+          WorkList.push_back(RegDef);
+      }
+    }
+  }
```

Comment:
I'm sorry for all the back and forth discussions, I think we misunderstood each other 😅 Generally, I think it isn't safe to traverse all uses. For our case, I believe it is only safe to traverse:
 - VADD.32 if one of its inputs eventually ends on a VBCST.32 of globals (or any mem location that is trackable)
 - or `VSEL.32` if both of its inputs eventually end on a VBCST.32 of globals (or any mem location that is trackable)

What do you think? Is this maybe too conservative?

Reply:
Discussed offline.